src/Abstractions/Entities/TaskEntity.cs

```diff
@@ -71,176 +69,51 @@ public interface ITaskEntity
-        {
-            await t;
-            return null;
-        }
+    /// <summary>
+    /// Gets a value indicating whether dispatching operations to <see cref="State"/> is allowed. State dispatch will
+    /// only be attempted if entity-level dispatch does not succeed. Default is <c>true</c>.
+    /// </summary>
+    protected bool AllowStateDispatch => true;
 
-        if (t.IsCompletedSuccessfully)
-        {
-            return default;
-        }
+    /// <summary>
+    /// Gets or sets the state for this entity.
+    /// </summary>
+    /// <remarks>
+    /// This will be hydrated as part of <see cref="RunAsync(TaskEntityOperation)"/>. The contents of this property
+    /// will be persisted to <see cref="TaskEntityContext.SetState(object?)"/> when the operation completes. Deleting
+    /// entity state can be accomplished by setting this to default(<typeparamref name="TState"/>).
+    /// </remarks>
+    protected TState State { get; set; } = default!; // leave null-checks to end implementation.
 
-        return new(Await(t));
-    }
+    /// <summary>
+    /// Gets the entity operation.
+    /// </summary>
+    protected TaskEntityOperation Operation { get; private set; } = null!;
 
-    static ValueTask<object?> AsGeneric(object result, Type type)
-    {
-        // result and type here must be some form of ValueTask<T>.
-        if ((bool)type.GetProperty("IsCompletedSuccessfully").GetValue(result))
-        {
-            return new(type.GetProperty("Result").GetValue(result));
-        }
-        else
-        {
-            Task t = (Task)type.GetMethod("AsTask", BindingFlags.Instance | BindingFlags.Public)
-                .Invoke(result, null);
-            return new(t.ToGeneric<object?>());
-        }
-    }
+    /// <summary>
+    /// Gets the entity context.
+    /// </summary>
+    protected TaskEntityContext Context => this.Operation.Context;
 
-    bool TryDispatchMethod(TaskEntityOperation operation, out object? result, out Type returnType)
+    /// <inheritdoc/>
+    public ValueTask<object?> RunAsync(TaskEntityOperation operation)
     {
-        Type t = this.GetType();
-
-        // Will throw AmbiguousMatchException if more than 1 overload for the method name exists.
-        MethodInfo? method = t.GetMethod(operation.Name, InstanceBindingFlags);
-        if (method is null)
-        {
-            result = null;
-            returnType = typeof(void);
-            return false;
-        }
-
-        ParameterInfo[] parameters = method.GetParameters();
-        object?[] inputs = new object[parameters.Length];
-
-        int i = 0;
-        ParameterInfo? inputResolved = null;
-        ParameterInfo? contextResolved = null;
-        ParameterInfo? operationResolved = null;
-        foreach (ParameterInfo parameter in parameters)
+        this.Operation = Check.NotNull(operation);
+        object? state = operation.Context.GetState(typeof(TState));
+        this.State = state is null ? default! : (TState)state;
```

Comment:
I think there is an important feature still missing here, regarding automatic first initialization of the entity state.

For most applications, the first time an entity is accessed, its state has to be somehow constructed/initialized. 

Of course one could write code for this in every operation, but that would be very inconvenient and error-prone for the user. With the design proposed here, users would be essentially forced to do this (or add their own extensions that do it). 

For reference, in the current DF interface we accommodate this "convenient API for initial creation" by

a. (for the function-based API) support specifying an initializer when calling GetState, and
b. (for the object-based API) automatically call the default constructor if the entity is null. 

I think we need something to support convenient initial creation in this API as well.

Reply:
Added a virtual method to `TaskEntity` to initialize state. This should cover the scenario.